### ISSUE TITLE
DALI Audio To Text Fails

### ISSUE THREAD

**Issue by agemagician** [NVIDIA/NeMo#2817]
Hello,

I am trying to train the conformer model with the DALI input pipeline, but it crash at the very early start.
Here is the logs:

```
[NeMo I 2021-09-14 16:32:36 mixins:147] Tokenizer SentencePieceTokenizer initialized with 128 tokens
[NeMo I 2021-09-14 16:32:36 ctc_bpe_models:172]
    Replacing placeholder number of classes (-1) with actual number of classes - 128
[NeMo W 2021-09-14 16:32:36 nemo_logging:349] /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/ops.py:627: DeprecationWarning: WARNING:
`normal_distribution` is now deprecated. Use `random.normal` instead.
      op_instances.append(_OperatorInstance(input_set, self, **kwargs))

Error executing job with overrides: ['+init_from_nemo_model=/home/ael/Scripts/nemo/training/pretrained_models/stt_en_conformer_ctc_small/stt_en_conformer_ctc_small.nemo']
Traceback (most recent call last):
  File "/home/ael/Scripts/nemo/training/speech_to_text_bpe.py", line 101, in <module>
    main()
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/core/config/hydra_runner.py", line 101, in wrapper
    _run_hydra(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/_internal/utils.py", line 377, in _run_hydra
    run_and_report(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/_internal/utils.py", line 214, in run_and_report
    raise ex
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/_internal/utils.py", line 378, in <lambda>
    lambda: hydra.run(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/_internal/hydra.py", line 111, in run
    _ = ret.return_value                                                                                                                                             File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/core/utils.py", line 233, in return_value
    raise self._return_value                                                                                                                                         File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)                                                                                                                       File "/home/ael/Scripts/nemo/training/speech_to_text_bpe.py", line 81, in main
    asr_model = EncDecCTCModelBPE(cfg=cfg.model, trainer=trainer)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/models/ctc_bpe_models.py", line 179, in __init__
    super().__init__(cfg=cfg, trainer=trainer)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/models/ctc_models.py", line 153, in __init__
    super().__init__(cfg=cfg, trainer=trainer)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/core/classes/modelPT.py", line 120, in __init__
    self.setup_training_data(self._cfg.train_ds)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/models/ctc_models.py", line 423, in setup_training_data
    self._train_dl = self._setup_dataloader_from_config(config=train_data_config)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/models/ctc_bpe_models.py", line 201, in _setup_dataloader_from_config
    dataset = audio_to_text_dataset.get_dali_bpe_dataset(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/data/audio_to_text_dataset.py", line 275, in get_dali_bpe_dataset
    dataset = audio_to_text_dali.AudioToBPEDALIDataset(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/data/audio_to_text_dali.py", line 645, in __init__
    super().__init__(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nemo/collections/asr/data/audio_to_text_dali.py", line 398, in __init__
    self._iter = DALIPytorchIterator(
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/plugin/pytorch.py", line 190, in __init__
    self._first_batch = DALIGenericIterator.__next__(self)
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/plugin/pytorch.py", line 201, in __next__
    outputs = self._get_outputs()
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/plugin/base_iterator.py", line 254, in _get_outputs
    outputs.append(p.share_outputs())
  File "/home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/pipeline.py", line 906, in share_outputs
    return self._pipe.ShareOutputs()
RuntimeError: Critical error in pipeline:
Error when executing GPU operator Pad encountered:
[/opt/dali/dali/operators/generic/pad.h:99] Assert on "a > 0" failed: Values of `align` argument must be positive.
Stacktrace (11 entries):
[frame 0]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali_operators.so(+0x55daee) [0x7ff58607aaee]
[frame 1]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali_operators.so(+0x138c8ed) [0x7ff586ea98ed]
[frame 2]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali_operators.so(+0x1360928) [0x7ff586e7d928]
[frame 3]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(void dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunHelper<dali::DeviceWorkspace>(dali::OpNode&, dali::DeviceWorkspace&)+0x845) [0x7ff584ff7745]
[frame 4]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunGPUImpl()+0x41c) [0x7ff584ff852c]
[frame 5]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunGPU()+0xe) [0x7ff584ff8f4e]
[frame 6]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(+0xa4ea8) [0x7ff584fa9ea8]
[frame 7]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(+0x123e8c) [0x7ff585028e8c]
[frame 8]: /home/ael/anaconda3/envs/nemo-1.3.0/lib/python3.9/site-packages/nvidia/dali/libdali.so(+0x70a55f) [0x7ff58560f55f]
[frame 9]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7ff6876676db]
[frame 10]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7ff68739071f]
```

The same model doesn't crash at this point when I don't use DALI.
The only difference is that I enable "use_dali: true" at the yaml file.

Any idea what is the cause of this problem and how I can solve it ?

**Comment by titu1994** [comment]
Is this with the main branch ?

**Comment by agemagician** [comment]
This is the latest release  "1.3.0"

**Comment by titu1994** [comment]
Can you try with the main branch, there was a bugfix for CPU + DALI execution which should land in the next NeMo release.

**Comment by agemagician** [comment]
Same error with the main branch.

**Comment by titu1994** [comment]
Could you share your config that builds the model without DALI?

**Comment by agemagician** [comment]
```
# It contains the default values for training a Conformer-CTC ASR model, large size (~120M) with CTC loss and sub-word encoding.

# Architecture and training config:
# Default learning parameters in this config are set for effective batch size of 2K. To train it with smaller effective
# batch sizes, you may need to re-tune the learning parameters or use higher accumulate_grad_batches.
# Here are the recommended configs for different variants of Conformer-CTC, other parameters are the same as in this config file.
# One extra layer (compared to original paper) is added to the medium and large variants to compensate for replacing the LSTM decoder with a linear one.
#
#  +-------------+---------+---------+----------+------------+-----+
#  | Model       | d_model | n_heads | n_layers | time_masks | lr  |
#  +=============+=========+========+===========+============+=====+
#  | Small  (13M)|   176   |    4   |    16     |     5      | 5.0 |
#  +-------------+---------+--------+-----------+------------+-----+
#  | Medium (30M)|   256   |    4   |    18     |     5      | 5.0 |
#  +-------------+---------+--------+-----------+------------+-----+
#  | Large (121M)|   512   |    8   |    18     |     10     | 2.0 |
#  +---------------------------------------------------------------+
#
# If you do not want to train with AMP, you may use weight decay of 0.0 or reduce the number of time maskings to 2
# with time_width=100. It may help when you want to train for fewer epochs and need faster convergence.
# With weight_decay=0.0, learning rate may need to get reduced to 2.0.

# You may find more info about Conformer-CTC here: https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/asr/models.html#conformer-ctc
# Pre-trained models of Conformer-CTC can be found here: https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/asr/results.html
# The checkpoint of the large model trained on LibriSpeech with this recipe can be found here: https://ngc.nvidia.com/catalog/models/nvidia:nemo:stt_en_conformer_ctc_large_ls

name: "Conformer-CTC-BPE-Small"

model:
  sample_rate: 16000
  log_prediction: true # enables logging sample predictions in the output during training
  ctc_reduction: 'mean_batch'

  train_ds:
    manifest_filepath: /mnt/local/extra/users/ael/dl/data/en/train_filtered_flac.json
    sample_rate: ${model.sample_rate}
    batch_size: 8 # you may increase batch_size if your memory allows
    shuffle: true
    num_workers: 4
    pin_memory: true
    use_start_end_token: false
    trim_silence: false
    max_duration: 17 # it is set for LibriSpeech, you may need to update it for your dataset
    min_duration: 0.1
    is_tarred: false
    #use_dali: true

  validation_ds:
    manifest_filepath:
    - /home/jho/nemo/github_ahmed/NeMo/tasks/en/weighted_combined1/datasets/librispeech_dev_clean.json
    - /home/jho/nemo/github_ahmed/NeMo/tasks/en/weighted_combined1/datasets/librispeech_dev_other.json
    sample_rate: ${model.sample_rate}
    batch_size: 8 # you may increase batch_size if your memory allows
    shuffle: false
    num_workers: 4
    pin_memory: true
    use_start_end_token: false
    is_tarred: false
    tarred_audio_filepaths: ''
    #use_dali: true

  test_ds:
    manifest_filepath:
    - /home/jho/nemo/github_ahmed/NeMo/tasks/en/weighted_combined1/datasets/librispeech_dev_clean.json
    - /home/jho/nemo/github_ahmed/NeMo/tasks/en/weighted_combined1/datasets/librispeech_dev_other.json
    sample_rate: ${model.sample_rate}
    batch_size: 8 # you may increase batch_size if your memory allows
    shuffle: false
    num_workers: 4
    pin_memory: true
    use_start_end_token: false
    is_tarred: false
    tarred_audio_filepaths: ''
    #use_dali: true

  # recommend small vocab size of 128 or 256 when using 4x sub-sampling
  # you may find more detail on how to train a tokenizer at: /scripts/tokenizers/process_asr_text_tokenizer.py
  tokenizer:
    dir: /home/ael/Scripts/nemo/training/pretrained_models/stt_en_conformer_ctc_small/  # path to directory which contains either tokenizer.model (bpe) or vocab.txt (wpe)
    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: ${model.sample_rate}
    normalize: "per_feature"
    window_size: 0.025
    window_stride: 0.01
    window: "hann"
    features: 80
    n_fft: 512
    log: true
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0
    pad_value: 0.0

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2 # set to zero to disable it
    # you may use lower time_masks for smaller models to have a faster convergence
    time_masks: 2 # set to zero to disable it
    freq_width: 27
    time_width: 0.05

  encoder:
    _target_: nemo.collections.asr.modules.ConformerEncoder
    feat_in: ${model.preprocessor.features}
    feat_out: -1 # you may set it if you need different output size other than the default d_model
    n_layers: 16
    d_model: 176

    # Sub-sampling params
    subsampling: striding # vggnet or striding, vggnet may give better results but needs more memory
    subsampling_factor: 4 # must be power of 2
    subsampling_conv_channels: -1 # -1 sets it to d_model

    # Feed forward module's params
    ff_expansion_factor: 4

    # Multi-headed Attention Module's params
    self_attention_model: rel_pos # rel_pos or abs_pos
    n_heads: 4 # may need to be lower for smaller d_models
    # [left, right] specifies the number of steps to be seen from left and right of each step in self-attention
    att_context_size: [-1, -1] # -1 means unlimited context
    xscaling: true # scales up the input embeddings by sqrt(d_model)
    untie_biases: true # unties the biases of the TransformerXL layers
    pos_emb_max_len: 5000

    # Convolution module's params
    conv_kernel_size: 31

    ### regularization
    dropout: 0.1 # The dropout used in most of the Conformer Modules
    dropout_emb: 0.0 # The dropout used for embeddings
    dropout_att: 0.1 # The dropout for multi-headed attention modules

  decoder:
    _target_: nemo.collections.asr.modules.ConvASRDecoder
    feat_in: null
    num_classes: -1
    vocabulary: []

  optim:
    name: adamw
    lr: 0.005
    # optimizer arguments
    betas: [0.9, 0.98]
    # less necessity for weight_decay as we already have large augmentations with SpecAug
    # you may need weight_decay for large models, stable AMP training, small datasets, or when lower augmentations are used
    # weight decay of 0.0 with lr of 2.0 also works fine
    weight_decay: 0.0

    # scheduler setup
    sched:
      name: NoamAnnealing
      d_model: ${model.encoder.d_model}
      # scheduler config override
      warmup_steps: null
      warmup_ratio: 0.05
      min_lr: 1e-6

trainer:
  gpus: -1 # number of GPUs, -1 would use all available GPUs
  num_nodes: 1
  max_epochs: 15
  max_steps: null # computed at runtime if not set
  val_check_interval: 0.5 # Set to 0.25 to check 4 times per epoch, or an int for number of iterations
  accelerator: ddp
  accumulate_grad_batches: 8
  gradient_clip_val: 0.0
  amp_level: O0 # O1/O2 for mixed precision
  precision: 32 # Should be set to 16 for O1 and O2 to enable the AMP.
  log_every_n_steps: 100  # Interval of logging.
  progress_bar_refresh_rate: 10
  resume_from_checkpoint: null # The path to a checkpoint file to continue the training, restores the whole state including the epoch, step, LR schedulers, apex, etc.
  num_sanity_val_steps: 1 # number of steps to perform validation steps for sanity check the validation process before starting the training, setting to 0 disables it
  check_val_every_n_epoch: 1 # number of evaluations on validation every n epochs
  sync_batchnorm: true
  checkpoint_callback: false  # Provided by exp_manager
  logger: false  # Provided by exp_manager
  plugins: ddp_sharded

exp_manager:
  exp_dir:  /mnt/local/extra/users/ael/dl/data/nemo/models/en/
  name: ${name}
  create_tensorboard_logger: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    # in case of multiple validation sets, first one is used
    monitor: "val_wer"
    mode: "min"
    save_top_k: 3

  # you need to set these two to True to continue the training
  resume_if_exists: true
  resume_ignore_no_checkpoint: true

  # You may use this section to create a W&B logger
  create_wandb_logger: true
  wandb_logger_kwargs:
    name: conformer-small-bpe-en-balanced-16k
    project: lumenvox-asr-en

hydra:
  run:
    dir: /mnt/local/extra/users/ael/dl/data/nemo/models/en/
```

**Comment by agemagician** [comment]
The above YAML file works without "use_dali: True". However, once it is enabled. The training crashes at the beginning as shown above.

**Comment by titu1994** [comment]
Can you try removing the "pad_to: 0" value for the preprocessor?

**Comment by agemagician** [comment]
Thanks. That made the training starts. However, I have notice two things:
1. The loss is much higher than without dali. With Dali around 150 and without Dali around 30.
I also noticed the training now is slower. Single epoch without Dali takes around 30 hours and with Dali takes around 38 hours.

**Comment by titu1994** [comment]
For (1) its reasonable because this was just a test to see if pad_to= 0 was causing the error. Without that flag, it defaults to pad_to=16 which damages the alignment of Conformer. I will send a bugfix later to address the case where pad_to = 0

For the speed I cant say why that's the case. We found the ASR processing speed to improve significantly when switching to DALI

**Comment by agemagician** [comment]
Great, hopefully, the fix will be released soon.

Regarding the second point, what is really interesting is that it is much faster during validation but slower during training.
I will retest it again after you release the bug fix.

Thanks a lot for your help.

**Comment by titu1994** [comment]
The bug fix should not have an effect on training speed, but lets see..

**Comment by agemagician** [comment]
Unfortunately, this fix doesn't solve the loss problem.
The loss for finetuning the conformer model by setting padding to 1 when using DALI is much higher than setting padding to 0 without DALI.
In my use case, the starting loss is about 150 with DALI and around 25 without DALI.

**Comment by titu1994** [comment]
Try setting padding to 0 for Dali. It should have no padding anymore.

**Comment by agemagician** [comment]
I did set it to 0, but the loss is starting around 150 with DALI compared to 25 without DALI.

From what I understood from the new fix:
`self.pad_to = max(params['pad_to'], 1) if 'pad_to' in params else 16`

That it will set the padding to 1 if it is zero.

**Comment by titu1994** [comment]
That's just used for align=(1, 1) below in Dali, which simply forces the shape to be a multiple of (1, 1) - ie any shape will be fine.

**Comment by agemagician** [comment]
Even before the training start, during the validation process on the validation dataset, I can see the logs from the prediction are much worse when using DALI compared to when I am not using it.

This is using the same YAML file with the same configuration except enabling and disabling DALI.

**Comment by agemagician** [comment]
Aha, thanks for the explination. Maybe, there is another problem in the DALI pipeline then.

**Comment by agemagician** [comment]
In this case, I will open a new issue, and I will try to create a Google Colab example that replicates the issue of different losses when using DALI.

**Comment by titu1994** [comment]
Hmm it's odd, Dalis audio processing should normally work similarly. I'll try looking at it but I'm not sufficiently familiar with Dali itself to resolve it quickly
